### PR TITLE
feat: api per tool plugin config

### LINKED ISSUE #3996: [FEATURE][PLUGINS]: Phase2 -  Plugin-Tool Policy Config CRUD API
Plugin-Tool Policy Config CRUD API

Design and implement a CRUD API that stores and enforces per-team, per-tool plugin configuration, enabling operators to apply plugin policies at the intersection of a team and a tool.

### 🧭 Type of Feature

Please select the most appropriate category:

- [ ] Enhancement to existing functionality
- [X] New feature or capability
- [ ] New MCP-compliant server
- [ ] New component or integration
- [ ] Developer tooling or test improvement
- [ ] Packaging, automation and deployment (ex: pypi, docker, quay.io, kubernetes, terraform)
- [ ] Other (please describe below)

### 🧭 Epic

**Title:** <High-level feature or capability>
**Goal:** What is the big-picture objective of this feature set?
**Why now:** Why is this needed? Who benefits?

---

### 🧑🏻‍💻 User Story 1

**As a:** platform administrator
**I want:** to create, read, update, and delete plugin policy configs scoped to a `(team_id, tool_id)` pair
**So that:** I can define which plugins run for a given team's tool invocations, and with what mode and config overrides

#### ✅ Acceptance Criteria
```gherkin
Scenario: Create a plugin policy config for a team/tool pair
  Given a valid team_id, tool_id, plugin name, mode, and config payload
  When I POST /admin/plugin-policies
  Then a new PluginToolPolicy record is created and returned with a unique id

Scenario: Read all plugin policy configs for a team
  Given a valid team_id
  When I GET /admin/plugin-policies?team_id={team_id}
  Then all policy records for that team are returned

Scenario: Update an existing policy config
  Given an existing plugin policy id
  When I PUT /admin/plugin-policies/{id} with new mode or config
  Then the record is updated and the new values are returned

Scenario: Delete a plugin policy config
  Given an existing plugin policy id
  When I DELETE /admin/plugin-policies/{id}
  Then the record is removed and no longer applied on tool invocation
```

---

### 📐 Design Sketch (optional)

Include a diagram, sketch, or flow (use [Mermaid](https://mermaid.js.org/) if desired):

### 🔗 MCP Standards Check

- [ ] Change adheres to current [MCP specifications](https://developer.ibm.com/tutorials/awb-handle-remote-tool-calling-model-context-protocol/)
- [ ] No breaking changes to existing MCP-compliant integrations
- [ ] If deviations exist, please describe them below:

---

### 🔄 Alternatives Considered

List any alternative designs, existing workarounds, or rejected ideas.

---

### 📓 Additional Context

Include related issues, links to discussions, issues, etc.

### PR DESCRIPTION
**Pull request by madhu-mohan-jaishankar**
## 🔗 Related Issue
Closes #3996

---

## 📝 Summary

Adds a `v1/tools/plugin_bindings` API that lets platform admins configure per-tool, per-team plugin policies in a single request. Each `(team_id, tool_name, plugin_id)` triple is upserted — existing rows are updated in place, new rows are inserted — making the endpoint idempotent.

**What's included:**

- **ORM model** (`ToolPluginBinding`) and **Alembic migration** for the new table
- **Pydantic schemas** with full validation: `ToolPluginBindingRequest`, `TeamPolicies`, `PluginPolicyItem`, and plugin-specific config schemas (`OutputLengthGuardConfig`, `RateLimiterConfig`, `SecretsDetectionConfig`) each with cross-field validators and `extra='forbid'`
- **Service layer** (`ToolPluginBindingService`): upsert, list (with optional team filter), delete
- **Router** (`/v1/tools/plugin_bindings`): POST, GET `/`, GET `/{team_id}`, DELETE `/{binding_id}`, wired to RBAC (`tools.manage_plugins`, `tools.read`)
- **66 tests** covering service and router with in-memory SQLite (no mocked DB or service responses), schema validation edge cases, and router error-handling wiring

---

## 🏷️ Type of Change
- [ ] Bug fix
- [x] Feature / Enhancement
- [ ] Documentation
- [ ] Refactor
- [ ] Chore (deps, CI, tooling)
- [ ] Other (describe below)

---

## 🧪 Verification

| Check                     | Command         | Status |
|---------------------------|-----------------|--------|
| Lint suite                | `make lint`     |        |
| Unit tests                | `make test`     |        |
| Coverage ≥ 80%            | `make coverage` |        |

---

## ✅ Checklist
- [ ] Code formatted (`make black isort pre-commit`)
- [x] Tests added/updated for changes
- [ ] Documentation updated (if applicable)
- [x] No secrets or credentials committed

---

## 📓 Notes (optional)

- Router catches only `ValueError` from the service layer (maps to 400). All other exceptions propagate as 500 — intentional, so server-side bugs are not silently swallowed as "bad request".

### CODE REVIEW

File: mcpgateway/routers/tool_plugin_bindings.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        HTTPException 403: If the caller lacks permission.
+
+    Examples:
+        >>> import asyncio
+        >>> asyncio.iscoroutinefunction(upsert_tool_plugin_bindings)
+        True
+    """
+    try:
+        caller_email: str = current_user_ctx["email"]
+        bindings = _service.upsert_bindings(db, request, caller_email=caller_email)
```
Comment (gcgoncalves) [inline mcpgateway/routers/tool_plugin_bindings.py:89]: A user with tools.manage_plugins can create bindings for any team, even teams they don't belong to.

Consider adding team membership validation, like so:

```python
user_teams = current_user_ctx.get("teams", [])
is_admin = current_user_ctx.get("is_admin", False)

if not is_admin:
    for team_id in request.teams.keys():
        if team_id not in user_teams:
            raise HTTPException(403, f"Not authorized for team {team_id}")
```

File: mcpgateway/services/tool_plugin_binding_service.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            List[ToolPluginBindingResponse]: All created/updated bindings, flattened.
+        """
+        results: List[ToolPluginBindingResponse] = []
+        now = utc_now()
+
+        for team_id, team_policies in request.teams.items():
+            for policy in team_policies.policies:
+                for tool_name in policy.tool_names:
+                    existing = (
+                        db.query(ToolPluginBinding)
```
Comment (gcgoncalves) [inline mcpgateway/services/tool_plugin_binding_service.py]: This might quickly become a performance issue. For instance, a request with 2 teams, 3 policies and 5 tools will run 30 (2 * 5 * 3) `SELECT` queries.

How about consolidating all existing bindings upfront, to then perform a single select?

```python
# Collect all (team_id, tool_name, plugin_id) tuples
keys = [(team_id, tool_name, policy.plugin_id.value) 
        for team_id, policies in request.teams.items()
        for policy in policies.policies
        for tool_name in policy.tool_names]

# Single query with IN clause
existing_map = {
    (b.team_id, b.tool_name, b.plugin_id): b
    for b in db.query(ToolPluginBinding).filter(
        tuple_(ToolPluginBinding.team_id, 
               ToolPluginBinding.tool_name,
               ToolPluginBinding.plugin_id).in_(keys)
    ).all()
}
```

**Review by gcgoncalves — CHANGES_REQUESTED** [review]

File: mcpgateway/routers/tool_plugin_bindings.py (hunk 1 above)
Comment (madhu-mohan-jaishankar) [reply mcpgateway/routers/tool_plugin_bindings.py:89]: Good catch, thanks. Added the team membership guard. The implementation collects all unauthorized teams before raising, so the caller gets a complete list in a single 403 rather than failing one team at a time.

File: mcpgateway/services/tool_plugin_binding_service.py (hunk 2 above)
Comment (madhu-mohan-jaishankar) [reply mcpgateway/services/tool_plugin_binding_service.py]: I went with team_id IN (team_ids) rather than tuple_().in_(keys) — SQLite (the default DB) doesn't support composite row-value expressions in IN, so tuple_ would break on SQLite while PostgreSQL handles it fine. The team_id IN approach fetches a slightly broader set (all bindings for the requested teams, not just the exact triples), but the result is still filtered in Python via dict lookup and the overhead is negligible in practice. Also consolidated the loop's db.flush() calls into a single flush after the loop.

Comment (gcgoncalves) [reply mcpgateway/services/tool_plugin_binding_service.py]: Nice. 👍

**Comment by madhu-mohan-jaishankar** [comment]
Closing this as the changes are raised in https://github.com/IBM/mcp-context-forge/pull/4068 along with plugin multi tenancy changes.
